walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,123 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    final Collection<String> requestAttendees = request.getAttendees();
+    final long requestDuration = request.getDuration();
+    ArrayList<TimeRange> eventList = createFlattenedEventList(events, requestAttendees);
+
+    if(requestDuration > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+    if(requestAttendees.isEmpty() || eventList.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    return findMeetings(eventList, requestDuration);
+  }
+  
+   private ArrayList<TimeRange> createEventList(Collection<Event> events, Collection<String> requestAttendees) {
+      ArrayList<TimeRange> eventList = new ArrayList<>();
+      for(Event event: events) {
+          Set<String> attendees = event.getAttendees();
+          for(String attendee: attendees) {
+              if(requestAttendees.contains(attendee)) {
+                eventList.add(event.getWhen());
+                break;
+              }
+          }
+      }
+      return eventList;
+  }
+
+  private ArrayList<TimeRange> createFlattenedEventList(Collection<Event> events, Collection<String> requestAttendees) {
+      ArrayList<TimeRange> eventList = createEventList(events, requestAttendees);
+      ArrayList<TimeRange> ret = new ArrayList<>();
+      Collections.sort(eventList, TimeRange.ORDER_BY_START);
+      int size = eventList.size();
+      if(size == 1) {
+          ret.add(eventList.get(0));
+          return ret;
+      }
+      for(int i = 0; i < size; i++) {
+          TimeRange curr = eventList.get(i);
+          TimeRange next = i + 1 == size ? null : eventList.get(i + 1);
+          if(next != null && curr.overlaps(next)) {
+              eventList.set(i + 1, TimeRange.fromStartEnd(curr.start(), Math.max(next.end(), curr.end()), false));
+          }
+          else {
+              ret.add(curr);
+          }
+      }
+      return ret;
+  }
+
+  private Collection<TimeRange> findMeetings(ArrayList<TimeRange> eventList, long requestDuration) {
+      ArrayList<TimeRange> ret = new ArrayList<>();
```

Comment:
Make this a more descriptive name, like `availableTimes` or `meetingTimes` or something